### DISCUSSION TITLE
What's the best practice to avoid Controller missing defaultValue warning.

### DISCUSSION THREAD

**Discussion by michael-land** [react-hook-form/react-hook-form#3653] [Q&A]
Hi all.

By design, Controller defaultValue should be a required prop, I used to coalesce undefined to null in my custom `Controller`.

```
 <Controller defaultValue={defaultValue || null} render={({value})=> <input value={value||""}/>}/>
```

However, most validation library treat null and undefined differently.

* `zod` and `superstruct` will raise error says null is not type string/number.
* `zod` and `superstruct` default value functionality only work if input is undefined.

I would like to know how you guys handle `Controller` default value. Do you end up writing custom validator?

**Comment by bluebill1049** [comment]
Hello @xiaoyu-tamu,

Let me put in some feedback on this one as well. I know this is annoying to see in the dev build, but there are many issues that were reported due to the user didn't support a default for the component. I will keep this comment updated and see if i can pull out some of the relevant issues that were reported.

**Reply by s97712** [reply]
In order to kill this warning, all fields need to have default values, which prevents me to use optional fields.

**Reply by bluebill1049** [reply]
what do you mean by optional fields? can you show me in a codesanbox? (it's a warning, i know it's annoying but it doesn't stop you from using it, but it may break which is case by case)

**Reply by s97712** [reply]
Optional field means allow to ignore the value or set to undefined,null
For example:
```tsx
<Controller rules={required: false} render={({value, onChange})=> <input value={value||""} onChange={onChange}/>}/>
```
It should be nothing wrong with this usage, but there are frequent warnings

**Reply by bluebill1049** [reply]
that's optional for rules, as an input itself it should still supply a `defaultValue`, it's not possible for us to predict what's in side render.
